io/src/read.cpp

```diff
@@ -0,0 +1,50 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Local include(s).
+#include "traccc/io/read.hpp"
+
+#include "traccc/io/read_cells.hpp"
+#include "traccc/io/read_digitization_config.hpp"
+#include "traccc/io/read_geometry.hpp"
+
+// OpenMP include(s).
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace traccc::io {
+
+demonstrator_input read(std::size_t events, std::string_view directory,
+                        std::string_view detector_file,
+                        std::string_view digi_config_file, data_format format,
+                        vecmem::host_memory_resource *mr) {
+
+    // Read in the detector configuration.
+    const geometry geom = io::read_geometry(detector_file);
+    const digitization_config digi_cfg =
+        io::read_digitization_config(digi_config_file);
+
+    // Construct the result object.
+    demonstrator_input result{events, mr};
+
+    // Read in the cell data for all events. In parallel if possible.
+#pragma omp parallel for
+    for (std::size_t event = 0; event < events; ++event) {
+        traccc::cell_container_types::host cells =
+            io::read_cells(event, directory, format, &geom, &digi_cfg, mr);
+
+        // Update the result container. (In a single thread.)
+#pragma omp critical
```

Comment:
Isn't `demonstrator_input` just a vector? Then this critical section has no use; every thread will be writing to a separate location in the vector anyway.

Reply:
:thinking: Very true. I just took this code as it was. But you're absolutely right.

Reply:
Ah sorry I did not realize that this was just old copy-pasted code. Could be worth changing though, might improve the performance a little. :smile:

Reply:
:+1: